_adr/89/index.adoc

```diff
@@ -0,0 +1,116 @@
+// \* No unexpanded acronyms
+// \* No undefined jargon
+
+// No need for a title heading, it's added by the template
+
+## Context and Problem Statement
+// What is the background against which this decision is being taken?
+
+The ADR intends to solve the same problem as the one described in https://architecture.appservices.tech/adr/5/[ADR-5: Kafka Cluster Certificate Generation].
+The proposed solution in that ADR only works for managed data plane clusters but it is not an acceptable solution for when users brings / registers their own data plane clusters.
+
+## Goals
+// Bulleted list of outcomes that this ADR, if accepted, should help achieve
+- Automated certificate management (generation, renewal, revocation) per data plane cluster
+- Avoid sharing same certificate between enterprise data plane clusters
+- Avoid exposing domain registrar & certificate issuers credentials onto enterprise data plane clusters
+
+## Non-goals
+// Bulleted list of outcomes that this ADR is not trying to achieve.
+- Automated certificate management per kafka instance.
+- Creating a pattern for and/or developing a central certificate management service
+
+## Current situation
+// Where are we now?
+
+The https://architecture.appservices.tech/adr/5/[ADR-5: Kafka Cluster Certificate Generation] describes the current situation where one certificate is shared across data planes. The certificate is renewed manually using the https://github.com/bf2fc6cc711aee1a0c2a/kas-sre-sops/blob/main/sops/key_management/control_plane/dataplane_certificate.asciidoc[data plane certificate SOP]
+
+## Proposal
+// What is the decision being proposed
+The proposed solution is going to use https://Letsencrypt.org[Letsencrypt] to automatically manage certificates.
+The solution consists of: 
+
+1. Generating one wildcard certificate per data plane cluster.
+For each new data plane, there will be a wildcard certificate for the `*.<cluster-id>.kafka.bf2.dev` subdomain where `<cluster-id>` is unique for each cluster.
+2. The fleet manager in its reconcilers will use https://github.com/caddyserver/certmagic[certmagic library] to generate certificate for each data plane and store them in a secure Storage. 
+When a kafka is assigned to a given data plane, all routes created will be under the `*.<cluster-id>.kafka.bf2.dev` domain. 
+For example, if the cluster id is `mycluster` then Kafka boostrap hosts will have the form of `foobar.mycluster.kafka.bf2.dev`.
+During Kafka reconciliation between the Fleetshard and the Fleet manager, certificates will be fetched from the secure storage by the fleet manager and they'll be passed down to the data plane via the existing certificate passing mechanism i.e via the ManagedKafka CR. 
+3. No migration to be performed for existing data plane.
+The proposed solution will be applied to only new data plane clusters.  
+Existing data plane clusters will continue to share the wildcard certificate on the `*.kafka.bf2.dev` domain.  
```

Comment:
Sounds good. I am also open to suggestions on whether & how we could consider migrating the existing data planes to using this solution. My suggestion was that we make this change a non disruptive one to existing users / Kafkas but I am open to alternative suggestion e.g go ahead with the change all over and offer a migration guide

Reply:
If we change the existing certificate, would that be disruptive? From what I understand it would force client to reconnect to the brokers, but other than that, are there any side effects?

Reply:
I inferred that we would need to change the Kafka URL because the Kafka broker URL is simething like `kafka-uuid-broker1.kafka.bf2.org`. We did not anticipate the need.

Reply:
meaning the need for separate subdomains per kafka deployments

Reply:
> I inferred that we would need to change the Kafka URL because the Kafka broker URL is simething like kafka-uuid-broker1.kafka.bf2.org. We did not anticipate the need.

That's the issue. 

> ah but if that's the case, why not ask for a multi named cert?

I've not investigate how this will look like but from face value it looks like we'll need to generate a certificate per kafka instance. This is not what's advertised here in the ADR, but this would solve the issue, however there are challenges that are related to this:
1. We'll like need to ask for overrrides of letsencrypt limits as generating per Kafka we are more likely to hit them often
2. In the near by future, we'll have Kafka expansion feature. Which may add some complexity to the certificate management as new brokers are scaled up / brokers are scaled down we'll need to make sure that the certificate knows of new brokers or doesn't know old brokers.